### PR TITLE
Fix code/test correspondence; explicitly skip test empty cells/tests.

### LINKED ISSUE #97: Testing a non-code cell confuses test collection
Tests and cells are not collected in the same way:

```python
def extract_cellsources(notebook):
    return [c['source'].split('\n') for c in notebook.cells if c.get('cell_type') == 'code']

def extract_celltests(notebook):
    return [c['metadata'].get('tests', []) for c in notebook.cells]
```

But they are later assumed to correspond:

```python
    # for cell of notebook,
    # assemble code to write
    for i, [code, test] in enumerate(zip(sources, tests)):
        # add celltest
        cells.append([i, [], 'def test_cell_%d(self):\n' % i])
```

Therefore, testing a non-code cell throws off the counts.

This is already apparent in the committed Untitled.ipynb notebook, where there is a test in a markdown cell, which then results in this syntax error in the generated Untitled_test.py script:

```
        assert df.equals(pd.DataFrame([{'1': 1, '2': 2}]))
        assert 1 == 1
        with patch('pandas.DataFrame.head') as mock:
            # Cell {4} content

            

        
        """)
```

This is masked by test failures being swallowed (#91).

### LINKED ISSUE #99: Generated test methods for empty cells are confusing
It's confusing, because they show up as a test, count towards coverage, and run all the previous cells.

Omit them? Skip test them? I probably favor skipping.

### PR DESCRIPTION
**Pull request by ceball**
Fixes #97 and fixes #99. Same tests as in #111, just updated for 1-based counting.

## Summary of new behavior

1. Test methods are only generated for code cells. If a non-code cell contains a test in its metadata, it's an error at generation time. Only code cells will be included in "cell coverage" calculation.

2. Test methods are named like `test_code_cell_n` (so you know it's only code cells), where `n` is the 1-based code cell number - i.e. they are numbered like execution count (if run from scratch).

Example: If there are 3 cells, and the 2nd cell is markdown, then there'll just be two test methods, `test_code_cell_1` and `test_code_cell_2`.

3. Code cells that are empty or do not have a test still generate a test method, but decorated with `@skip(reason)` (where `reason` is `"empty code cell"` or `"cell has no test"`). 

4. Tests that do not contain a cell injection generate a skipped test method (`reason` is `"no cell injection"`).

## Changes in this PR

* Fix correspondence of code and test cells (e.g. markdown caused them to go out of sync)

* Change cell/test numbering to start from 1, and only include code cells (matching other notebook tools).

* If a non-code cell contains test in metadata, it's an error during test script generation.

* Create explicitly skipped tests for:
    * empty cell
    * empty test
    * code cell not injected into test

To do:
  * [x] Fix windows tests

Future work:
  * Update coverage calculation (see https://github.com/jpmorganchase/nbcelltests/pull/112#issuecomment-630909164)

EDIT: clarified that coverage will be addressed in a subsequent PR once behavior here is agreed.

### CODE REVIEW

**Comment by vidartf** [comment]
I see there is still an unchecked todo here. Is this ready for review?

**Comment by vidartf** [comment]
Also: Does this supersede #111 then?

**Comment by ceball** [comment]
#111 just demonstrates the same tests as are here failing on master. #111 is not for merging; it's for information only.

> I see there is still an unchecked todo here. Is this ready for review?

@vidartf Yes, it's ready for review. I have removed the to-do. Since the coverage calculation* is not altered here, and is not tested, I should be allowed to address that in a future PR I think - _after_ the behavior of testing here is agreed. The behavior changes here imply that the coverage calculation should change. It's currently:
```
num code cells with %cell in their tests / num code cells
```
This PR being accepted implies coverage should change to:
```
num non-empty code cells with %cell in their tests / num non-empty code cells
```
Note that the current definition of coverage can be gamed by adding empty code cells with tests :)

(*) Actually there are three separate calculations of coverage in the code. Two of them are the same, one is broken. Still, none of those three calculations or components thereof are touched by this PR.

File: nbcelltests/tests/test_test.py [hunk 1]
```diff
@@ -60,28 +60,111 @@ def _import_from_path(path, module_name):
...
+        else:
+            raise Exception("Test script should fail to generate")
+
+
+class TestSkips(_TestCellTests):
+    """Tests that various conditions result in skipped tests"""
+
+    NBNAME = SKIPS
+
+    # the tests are independent, so it's fine to call setUpClass and
```
Comment (vidartf) [inline nbcelltests/tests/test_test.py:128]: It is fine to call it, but we might be missing some conditions that are only triggered if it is not. Ok since you mention a way to resolve in the future 👍

File: nbcelltests/test.py [hunk 2]
```diff
@@ -12,41 +12,81 @@
...
+# throughout the code.
+
 INDENT = '    '
 
 
-def assemble_code(sources, tests):
-    cells = []
+def _is_empty(source):
+    # TODO: py2 utf8
+    return len(ast.parse(source).body) == 0
```
Comment (ceball) [inline nbcelltests/test.py:33]: Apart from the TODO already noted, this also would not handle magics - but it should.

File: nbcelltests/tests/test_test.py [hunk 3]
```diff
@@ -60,28 +60,111 @@ def _import_from_path(path, module_name):
...
-        tf = tempfile.NamedTemporaryFile(mode='w', suffix='.py', delete=False, encoding='utf8')
-        tf_name = tf.name
+        assert hasattr(cls, "NBNAME"), "Subclasses must have NBNAME attribute."  # TODO: make actually abstract
+        cls.generated_tests = _generate_test_module(notebook=cls.NBNAME, module_name="nbcelltests.tests.%s.%s" % (__name__, cls.__name__))
+
+    def _assert_skipped(self, mthd, reason):
+        # TODO: actually calling the skipped method here ends
+        # everything! It's some pytest issue, I think. So for now we
+        # are checking that it would be skipped, rather than that it is
+        # definitely skipped.
```
Comment (vidartf) [inline nbcelltests/tests/test_test.py:97]: We should probably make an issue to track this if we merge as-is.

File: nbcelltests/test.py [hunk 4]
```diff
@@ -12,41 +12,81 @@
...
+import ast
 from .define import TestMessage, TestType
-from .shared import extract_cellsources, extract_celltests, extract_extrametadata
+from .shared import extract_extrametadata
 from .tests_vendored import BASE, JSON_CONFD
 
+# TODO: eventually want assemble() and the rest to be doing something
+# better than building up code in strings. It's tricky to work with,
+# tricky to read, can't lint, can't test easily, unlikely to be
+# generally correct, etc etc :)
```
Comment (vidartf) [inline nbcelltests/test.py:23]: It is indeed hard to follow the code in the review ;)

**Comment by vidartf** [comment]
> This PR being accepted implies coverage should change to: ...

I would say the final goal would be to get proper line/branch coverage reports :)

File: nbcelltests/shared.py [hunk 5]
```diff
@@ -113,6 +105,7 @@ def extract_extrametadata(notebook, override=None):
     base['cell_lines'] = []
 
     for c in notebook.cells:
+        # TODO if not c['cell_type'] == 'source'
```
[the remark this answers is not in the document]
Comment (ceball) [reply nbcelltests/shared.py:108]: Actually, ha ha, it's `code` not `source`, but the point stands :) 

I want to clean up various repeated/inconsistent bits like this in a future PR...

File: nbcelltests/test.py [hunk 6]
```diff
@@ -12,41 +12,81 @@
...
+        if cell.get('cell_type') != 'code':
+            if len(test_lines) > 0:
+                raise ValueError("Cell %d is not a code cell, but metadata contains test code!" % i)
+            continue
+
+        code_cell += 1
+
+        if _is_empty(cell['source']):
+
+            skiptest = "@nbcelltests.tests_vendored.unittest.skip('empty code cell')\n" + INDENT
```
[the remark this answers is not in the document]
Comment (ceball) [reply nbcelltests/test.py:62]: Mentioned in a comment at the start, just above: the only import in the template in tests_vendored is nbcelltests. Adding a unittest import to the template would be an unused name in the case there are no skipped tests. Adding a unittest import if unittest ends up being used in the generated tests would be possible, but also awkward. So I say easier to live with this awkward way until the code is no longer built up in a big string like this. (Which I am proposing is cleanup for a future PR.)

File: nbcelltests/tests/test_test.py (hunk 3 above)
Comment (ceball) [reply nbcelltests/tests/test_test.py:97]: Will make an issue to track this if this PR is merged with this in. 

It could be time consuming to track down (I couldn't track it down in the 5 minutes I tried...); I also feel like we might end up with an overall better way of doing things that obviates this.

File: nbcelltests/test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vidartf) [reply nbcelltests/test.py:62]: I would say this breaks modularization, and therefore is a code smell, but happy to leave for a larger cleanup later 👍

**Comment by ceball** [comment]
>> This PR being accepted implies coverage should change to: ...
>
> I would say the final goal would be to get proper line/branch coverage reports :)

Ok, I should have said the "cell tests coverage" or similar, to distinguish from actual code coverage (see also #109).

File: nbcelltests/test.py (hunk 2 above)
Comment (vidartf) [reply nbcelltests/test.py:33]: If it has a magic it won't be empty, so probably wrap in a try/except block?

**Comment by ceball** [comment]
Thanks, I'll follow up what we discussed here (e.g. creating an issue to track pytest in pytest skip handling) and then merge once I've done that.

File: nbcelltests/tests/test_test.py (hunk 3 above)
Comment (ceball) [reply nbcelltests/tests/test_test.py:97]: #117

File: nbcelltests/tests/test_test.py (hunk 1 above)
Comment (ceball) [reply nbcelltests/tests/test_test.py:128]: This can be improved when we generate test notebooks. But ideally we wouldn't be invoking the setup and test methods manually in the first place, I guess.

File: nbcelltests/test.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ceball) [reply nbcelltests/test.py:62]: But this is just a string - it's not code! Therefore there cannot be a code smell ;)

File: nbcelltests/test.py (hunk 4 above)
Comment (ceball) [reply nbcelltests/test.py:23]: Although I added the TODO, this is not part of the current PR (was like this before the PR, and still like it after).

**Comment by ceball** [comment]
Hmm, azure doesn't seem to be working here any more?
